padertorch/train/trainer.py

```diff
@@ -257,57 +267,148 @@ def train(
                 # TypeError: object of type '...' has no len()
                 max_it_len = None
             hooks.append(ProgressBarHook(self._stop_trigger, max_it_len))
         hooks = sorted(hooks, key=lambda h: h.priority, reverse=True)
 
+        if len(device) >= 2:
+            import textwrap
+            print(
+                'WARNING: You called padertorch.Trainer.train with multiple\n'
+                +
+                textwrap.indent(
+                    'devices. With this the trainer will use data parallel to\n'
+                    'utilize the multiple GPUs to speedup your training.\n'
+                    'We observed some problems with some versions of pytorch.\n'
+                    'In 1.4 the performance on a NN was quite bad and accoring to\n'
+                    'https://github.com/pytorch/pytorch/issues/33552\n'
+                    'this was because the RNNs get no gradients.\n'
+                    'In 1.5 the training got stuck, the reason is unclear in the'
+                    'moment.\n'
+                    'With Pytorch <= 1.3 we have not tested the code.\n'
+                    f'Your pytorch version is: {torch.__version__}',
+                    ' ' * len('WARNING: ')
+                )
+            )
+
+        assert self.virtual_minibatch_size % len(device) == 0, (self.virtual_minibatch_size, device)
+        assert len(device) > 0, (self.virtual_minibatch_size, device)
+
         # ================ MAIN TRAINING LOOP! ===================
         try:
-            # Count epochs up to infinity if not any stop condition is met. A
-            # typical stop condition is a firing `StopTrainingHook`.
-            for self.epoch in itertools.count(start=self.epoch):
-                epoch_start = True
-                for hook in hooks:
-                    hook.pre_step(self)
-
-                for self.iteration, example in self.train_timer(
-                    key='time_per_data_loading',
-                    iterable=enumerate(
-                        train_iterator,
-                        start=self.iteration,
-                    )
-                ):
-                    if epoch_start:
-                        epoch_start = False
-                    else:
-                        for hook in hooks:
-                            hook.pre_step(self)
-                    with self.train_timer['time_per_step']:
-                        model_output, review = self.train_step(
-                            example,
-                            optimize=(self.iteration+1) % self.virtual_minibatch_size == 0,
-                        )
+            train_iterable = None
+            while True:
+                new_epoch = False
+                if train_iterable is None:
+                    new_epoch = True
 
+                    # Call pre_step between the epochs.
+                    # We call it here, so it is done, before the iteration
+                    # over the train_dataset starts.
```

Comment:
In general, it is not necessary.
The reason is, that this can reduce the memory footprint.
When you trigger the validation after an epoch the train dataset shouldn't start to load the next examples.